### PR TITLE
Use standard C or C++ atomics for memoryviews

### PR DESCRIPTION
**Pull request by da-woods**
Use them in preference to the existing compiler-specific
atomic types.

In both cases this requires the 2011 edition of the standard.

Closes https://github.com/cython/cython/issues/4923

### CODE REVIEW

**Comment by scoder** [comment]
Looks good. I'll wait until after the next alpha release with a merge since I'd like to have a safe 3.0 version out there that supports Python 3.11, and this addition seems somewhat risky.

**Comment by da-woods** [comment]
> Looks good. I'll wait until after the next alpha release with a merge since I'd like to have a safe 3.0 version out there that supports Python 3.11, and this addition seems somewhat risky.

Yeah that seems reasonable. I haven't confirmed that it works on MSVC yet which I think is the main risk of breaking something.

File: Cython/Utility/MemoryView_C.c
Comment (da-woods) [reply Cython/Utility/MemoryView_C.c]: Thanks yes, copy-paste mistake

File: Cython/Utility/MemoryView_C.c
Comment (da-woods) [reply Cython/Utility/MemoryView_C.c:39]: I've added a comment.

I'm inclined not to get into using `<stdatomic.h>` in C++ mode. Everything I've seen suggests that it's very compiler-specific and inconsistent. And I'm trying to get away from compiler-specific and inconsistent as much as possible here

Comment (da-woods) [reply Cython/Utility/MemoryView_C.c:39]: > For reference, even VS 2017 (another "17", yay!) is already EOL, so as soon as Python 2.7 is dropped (and thus VS2008), this could be made unconditional. Surely Cython will want to have conservative lower bounds for compiler versions, but older than VS2017 (or maybe even VS2019) makes no sense IMO.

Possibly. There is a non-atomic fallback of course (and it looks like it's been used for years on Windows) so in this case Cython could support any users choosing to use really odd/old compilers with the fallback.

> In case it's of interest, this is the [page](https://scipy.github.io/devdocs/dev/toolchain.html) where scipy documents the relevant constraints for itself (that I help maintain) - as of 1.9, it requires VS2019 (with the vc142 toolchain).

That was useful. The slight difference is that the majority of Scipy users probably download pre-compiled versions while compiling your own code is the core feature of Cython. So it needs to be a little more tolerant.

File: Cython/Utility/MemoryView_C.c
Comment (scoder) [inline Cython/Utility/MemoryView_C.c:50]: I don't see much value in leaving these in the generated C files, but I'll leave the decision to you.
However, I do prefer honouring the readers of my comments by presenting them complete sentences and correct punctuation wherever it seems appropriate. I feel that this applies here. :)
```suggestion
    // C11 atomics are available.
    // Require ATOMIC_INT_LOCK_FREE because I'm nervous about the __pyx_atomic_int[2]
    // alignment trick in MemoryView.pyx if it uses mutexes.
```

File: Cython/Utility/MemoryView_C.c
Comment (scoder) [inline Cython/Utility/MemoryView_C.c:69]:
```suggestion
    // C++11 atomics are available.
    // Require ATOMIC_INT_LOCK_FREE because I'm nervous about the __pyx_atomic_int[2]
    // alignment trick in MemoryView.pyx if it uses mutexes.
```

File: Cython/Utility/MemoryView_C.c
Comment (scoder) [inline Cython/Utility/MemoryView_C.c:32]:
```suggestion
// For standard C/C++ atomics, get the headers first so we have ATOMIC_INT_LOCK_FREE
// defined when we decide to use them.
```

**Review by scoder — APPROVED** [review]

File: Cython/Utility/MemoryView_C.c
Comment (da-woods) [reply Cython/Utility/MemoryView_C.c]: I'm not sure it was a deliberate decision to leave these in the C files.

**Pull request merged**
